### ISSUE TITLE
IndexOutOfRangeException when trying to Emit/GetDiagnostics from a Compilation

### ISSUE THREAD

**Issue by gerich-home** [dotnet/roslyn#1009]
Hi,

I am playing with Roslyn API and as an experiment I build a simple AOP framework that rewrites methods decorated with aspect attributes in the existing syntax tree in order to call aspects on method boundaries.
When I was trying to Emit assembly from the rewritten Syntax Tree I get an uncaught IndexOutOfRangeException (Index was outside the bounds of the array.)
Stack Trace:
```
   at Microsoft.CodeAnalysis.CSharp.Binder.ConvertAndBindArrayInitialization(DiagnosticBag diagnostics, InitializerExpressionSyntax node, ArrayTypeSymbol type, Nullable`1[] knownSizes, Int32 dimension, ImmutableArray`1 boundInitExpr, Int32& boundInitExprIndex)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindArrayInitializerList(DiagnosticBag diagnostics, InitializerExpressionSyntax node, ArrayTypeSymbol type, Nullable`1[] knownSizes, Int32 dimension, ImmutableArray`1 boundInitExprOpt)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindArrayCreationWithInitializer(DiagnosticBag diagnostics, ExpressionSyntax creationSyntax, InitializerExpressionSyntax initSyntax, ArrayTypeSymbol type, ImmutableArray`1 sizes, ImmutableArray`1 boundInitExprOpt)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindArrayCreationExpression(ArrayCreationExpressionSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionAllowArgList(ExpressionSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindValueAllowArgList(ExpressionSyntax node, DiagnosticBag diagnostics, BindValueKind valueKind)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindArgumentAndName(AnalyzedArguments result, DiagnosticBag diagnostics, Boolean hadError, CSharpSyntaxNode argumentSyntax, ExpressionSyntax argumentExpression, NameColonSyntax nameColonSyntax, RefKind refKind, Boolean allowArglist)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindArgumentAndName(AnalyzedArguments result, DiagnosticBag diagnostics, Boolean hadError, ArgumentSyntax argumentSyntax, Boolean allowArglist)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindArgumentsAndNames(SeparatedSyntaxList`1 arguments, DiagnosticBag diagnostics, AnalyzedArguments result, Boolean allowArglist)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindArgumentsAndNames(ArgumentListSyntax argumentListOpt, DiagnosticBag diagnostics, AnalyzedArguments result, Boolean allowArglist)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindInvocationExpression(InvocationExpressionSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindValue(ExpressionSyntax node, DiagnosticBag diagnostics, BindValueKind valueKind)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionStatement(CSharpSyntaxNode node, ExpressionSyntax syntax, Boolean allowsAnyExpression, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionStatement(ExpressionStatementSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindStatement(StatementSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindBlock(BlockSyntax node, DiagnosticBag diagnostics, Binder blockBinder)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindBlock(BlockSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.BindMethodBody(MethodSymbol method, TypeCompilationState compilationState, DiagnosticBag diagnostics, Boolean generateDebugInfo, ConsList`1& debugImports)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol symbol)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass4.<CompileNamedTypeAsTask>b__5()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.WaitForWorkers()
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethodBodies(CSharpCompilation compilation, PEModuleBuilder moduleBeingBuiltOpt, Boolean generateDebugInfo, Boolean hasDeclarationErrors, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileImpl(CommonPEModuleBuilder moduleBuilder, Stream win32Resources, Stream xmlDocStream, Boolean generateDebugInfo, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Compilation.Compile(CommonPEModuleBuilder moduleBuilder, Stream win32Resources, Stream xmlDocStream, Boolean generateDebugInfo, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, CompilationTestData testData, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.FileSystemExtensions.Emit(Compilation compilation, String outputPath, String pdbPath, String xmlDocPath, String win32ResourcesPath, IEnumerable`1 manifestResources, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpFileSystemExtensions.Emit(CSharpCompilation compilation, String outputPath, String pdbPath, String xmlDocumentationPath, String win32ResourcesPath, IEnumerable`1 manifestResources, CancellationToken cancellationToken)
   at Roslyn.AOP.Tests.AspectsWeaverTests.EmitAspectedCompilation() in c:\Users\sgerasimov\Documents\Visual Studio 2012\Projects\RoslynAOP\Roslyn.AOP.Tests\AspectsWeaverTests.cs:line 89
   at Roslyn.AOP.Tests.AspectsWeaverTests.AssemblyGeneratedWithRewriteWithAspectsUsesTestAspect_FailsWithRuntimeError() in c:\Users\sgerasimov\Documents\Visual Studio 2012\Projects\RoslynAOP\Roslyn.AOP.Tests\AspectsWeaverTests.cs:line 27
```

Steps to reproduce:
1. Open solution from file: https://www.dropbox.com/s/kypsb2kz41dsf8f/RoslynAOP.zip?dl=0
2. Run tests
3. Catch an exception in AssemblyGeneratedWithRewriteWithAspectsUsesTestAspect_FailsWithRuntimeError test:
```
            LoadOriginalCompilation();

            _aspectedTree = _aspectsWeaver.RewriteWithAspects();

            // Fails:
            EmitAspectedCompilation();

            VerifyAspectsWereAppliedToCode();
```

If you try to run AssemblyGeneratedWithRewriteWithAspectsUsesTestAspect_HappyScenario test, then you will see the transormed source code is parsed into a fresh Syntax Tree and here it works:

```
            LoadOriginalCompilation();

            _aspectedTree = _aspectsWeaver.RewriteWithAspects();

            // Parse new tree from aspected source code again, insteas of using composed one:
            _aspectedTree = CSharpSyntaxTree.ParseText(_aspectedTree.ToString());

            // Works now:
            EmitAspectedCompilation();

            VerifyAspectsWereAppliedToCode();
```

**Comment by gerich-home** [comment]
I've continued my investigations, and I tried to compare Syntax trees:
```
            _aspectedTree = _aspectsWeaver.RewriteWithAspects();
            var a = _aspectedTree.GetRoot()
                .DescendantNodes().OfType<ArrayCreationExpressionSyntax>().First()
                .DescendantNodes().OfType<ArrayRankSpecifierSyntax>().First();

            // Parse new tree from aspected source code again, insteas of using composed one:
            _aspectedTree = CSharpSyntaxTree.ParseText(_aspectedTree.ToString());
            var b = _aspectedTree.GetRoot()
                .DescendantNodes().OfType<ArrayCreationExpressionSyntax>().First()
                .DescendantNodes().OfType<ArrayRankSpecifierSyntax>().First();
```

If you compare these trees by yourself, you will see that `a.Rank` equals `0` and `b.Rank` equals `1`.
I've compared the contents of `a.Sizes` and `b.Sizes` and obeserved that `b` contains single child node of type `OmittedArraySizeExpressionSyntax`.

So the problem here is that when you create array type with the following construct:
```
            var objectArrayType = SyntaxFactory.ArrayType(
                SyntaxFactory.PredefinedType(SyntaxFactory.Token(SyntaxKind.ObjectKeyword)),
                SyntaxFactory.List(new[] { SyntaxFactory.ArrayRankSpecifier() })
                );
```
then `ArrayRankSpecifier` gets constructed with no sizes, but should have single `OmittedArraySizeExpressionSyntax` child node.
You can always use smth like that as a workaround:
```
            var objectArrayType = SyntaxFactory.ArrayType(
                SyntaxFactory.PredefinedType(SyntaxFactory.Token(SyntaxKind.ObjectKeyword)),
                SyntaxFactory.List(new[]
                {
                    SyntaxFactory.ArrayRankSpecifier(
                        SyntaxFactory.SeparatedList<ExpressionSyntax>(new[] {SyntaxFactory.OmittedArraySizeExpression()}))
                })
                );
```

So to summarize there are 2 issues:
1. `SyntaxFactory.ArrayRankSpecifier()` creates an **invalid** node with no `OmittedArraySizeExpression` child (not sure if it is the issue, because it can be stringified to the correct source).
2. `Emit` fails with **uncaught exception** when invalid ArrayRankSpecifier exists.

**Comment by gafter** [comment]
The Roslyn compilers do not support compilation from hand-built trees, because such trees are typically not in a form that could have been built by the parser for a correct (or even incorrect) program. To diagnose the problems with your hand-built tree, serialize it into a string and then run the parser on the result. Do you get back an identical tree? If not, you've likely identified the problem.

**Comment by gerich-home** [comment]
As I mentioned, I serialized, parsed and compared resulting tree with orinal one and they are different.
According to your words it is likely a problem.

Please explain if my understanding is wrong.

**Comment by gafter** [comment]
@gerich-home I think we agree that was the problem.

**Comment by gerich-home** [comment]
In such case I do not understand why no Bug label is applied to this issue, why it is still just a Question?

**Comment by gafter** [comment]
Because it is not a bug. Roslyn is not *supposed* to be fed syntax trees that could not have been produced by the parser.
